### DISCUSSION TITLE
AWG

### DISCUSSION THREAD

**Discussion by Jingfu2000** [Ulm-IQO/qudi-core#102] [Q&A]
In the current version for Qudi, we generate the whole pulse sequnece in one single waveform, e.g., for 100 points in T1 measurement.  There are surely some advantages. However, there are also problems from the limitation of AWG memory size.  Could we generate one waveform for one point, i..e, for one delay time? In this way, we can measure much more points for one curve.  
Thanks.

**Comment by NicolasStaudenmaier** [comment]
Hi Jingfu, pulsed measurements in qudi-iqo-modules allow to use the sequence mode of an AWG. With this single "waveforms" are created (termed `PulseBlockEnsembles`). Playing a single waveform is the standard way. In sequence mode, however, the PulseBlockEnsembles can be composed to a sequence by playing an arbitrary list of waveforms (possibly with repetitions), what is then called a `PulseSequence`. 

For T1 measurements in particular, note that there are three different generation methods in the [basic_predefined_methods](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/main/src/qudi/logic/pulsed/predefined_generate_methods/basic_predefined_methods.py) of the qudi-iqo-modules. One creating an ensemble with equally spaced points (line 637 in the file) and one creating an ensemble but with logarithmically spaced points (line 697). Logarithmic spacing has the advantage that more points for shorter times are measured where more absolute decay is expected. The third one (line 1003) writes a PulseSequence with logarithmically spaced points. Here AWG memory is kept very low by using a single idle-ensemble that is repeated integer k times for different data points where k is exponentially growing. 

It is possible to create any own, custom `PulseSequence`. The existing one for the T1 measurement can be used as blueprint. In your own generation you have create the PulseBlockEnsembles first and then append them to the sequence.

**Reply by Jingfu2000** [reply]
Hi Nicolas,
Thanks a lot! 
best,
Jingfu

**Comment by timoML** [comment]
Please note that sequence mode works well for your issue as long as you do not care about phases between sequence steps (e.g. MW pulses). If you spread MW pulses over multiple steps, jitter from the AWG sequencer becomes a major concern. Depending on your device specs, this might or might not work.

**Comment by Jingfu2000** [comment]
Thanks. 
In T1 measurement, e.g., I can finish the measurement for one point at certain delay using one waveform.  In my previos lab, this result of the count rate is saved in a buffer of NI-DAQ counter. Then the above precedure is repeated for 100k, e.g., times. Then the computer read the buffer from NI-DAQ.  Then the sofeware goes to next delay time. Do we also use NI-DQA counters in Qudi?

**Reply by timoML** [reply]
The pulsed toolchain is not implemented for such sequential acqusition, as it's much more susceptible to low frequency drifts. We use fast counters (with bandwidths much higher than a nidaq) to acquire all taus in one shot. Some people might have tried implementing sequential acquisition for special purposes (maybe ask @IliaChup).

**Reply by Jingfu2000** [reply]
thanks.

**Reply by IliaChup** [reply]
I am working with a similar approach using NiDAQ. But it is not implemented in Qudi. I am just doing gated data acquisition on NiDAQ programmed on LabView. Then alternating sequences on AWG by triggers in order to synchronize measurements.

**Comment by Jingfu2000** [comment]
thanks Ilia. 
Labview and NiDAQ can always work well together.  The AWG should be programmed by Qudi. Can we use Qudi to call the functions compiled by  labview ? 
Best,
Jingfu
